ufl/formoperators.py

```diff
@@ -139,11 +151,28 @@ def adjoint(form, reordered_arguments=None):
 
     If the form is a baseform instance instead of a Form object, we return an Adjoint
     object instructing the adjoint to be computed at a later point.
+
+    When `adjoint` is being called multiple times on the same form, expanding derivatives
+    become expensive -> `derivatives_expanded` enables to use caching mechanisms to avoid that.
     """
     form = as_form(form)
-    if isinstance(form, Form):
-        form = expand_derivatives(form)
-        return compute_form_adjoint(form, reordered_arguments)
+    if isinstance(form, BaseForm):
+        # Allow BaseForm objects that are not BaseForm such as Adjoint since there are cases
+        # where we need to expand derivatives: e.g. to get the number of arguments
+        #   => For example: Adjoint(Action(2-form, derivative(u,u)))
+        try:
+            if not derivatives_expanded:
+                # For external operators differentiation may turn a Form into a FormSum
+                form = expand_derivatives(form)
+            if isinstance(form, Form):
+                return compute_form_adjoint(form, reordered_arguments)
+        except NotImplementedError:
+            # Catch cases where expand derivatives is not implemented
+            # e.g. `adjoint(Adjoint(M))` where M is a ufl.Matrix,
+            # expand_derivatives(M) will be M (no derivatives taken)
+            # and expand derivatives of Adjoint only works if when we push it through the Adjoint
+            # we get 0.
+            pass
```

Comment:
Remove since `Adjoint` differentiation, which is the reason motivating that bit, is not supported anymore